sstables/storage.cc

```diff
@@ -55,8 +88,13 @@ class filesystem_storage final : public sstables::storage {
     future<> move(const sstable& sst, sstring new_dir, generation_type generation, delayed_commit_changes* delay) override;
     future<> rename_new_file(const sstable& sst, sstring from_name, sstring to_name) const;
 
-    virtual void change_dir_for_test(sstring nd) override {
-        _dir = nd;
+    future<> change_dir(sstring new_dir) {
+        auto old_dir = std::exchange(_dir, opened_directory(new_dir));
+        return old_dir.close();
```

Comment:
nice

Reply:
I'm torn if this needs to be a method of open_directory, like
```c++
future<> chdir(std::filesystem::path new_path) {
    _pathname = std::move(new_path);
    auto f = std::move(_file);
    return f ? f.close() : make_ready_future();
}
```

Reply:
We don't need to make opened_directory a heavyweight abstraction that completely hides the details. It's a helper and can be kept lightweight.

Reply:
@bhalevy 

> nice

Thanks; credit for the [original idea](https://github.com/scylladb/scylladb/pull/19624#pullrequestreview-2173477444) is due @raphaelsc